### PR TITLE
Enable worker pool support in the CICD recipe

### PR DESCRIPTION
**Pull request by pasha-gh**

### CODE REVIEW

File: templates/tfengine/components/cicd/triggers.tf [hunk 1]
```diff
@@ -64,6 +69,7 @@ resource "google_cloudbuild_trigger" "validate_{{.name}}" {
   substitutions = {
     _TERRAFORM_ROOT = "{{$terraform_root}}"
     _MANAGED_DIRS = "{{$managed_dirs}}"
+    _WORKER_POOL = "{{$worker_pool}}"
```
Comment (xingao267) [inline templates/tfengine/components/cicd/triggers.tf:72]: we should do this for other triggers as well

File: templates/tfengine/components/cicd/triggers.tf [hunk 2]
```diff
@@ -23,6 +23,11 @@
   {{- $terraform_root_prefix = ""}}
 {{- end}}
 
+{{- $worker_pool := ""}}
+{{- if has . "worker_pool"}}
+  {{- $worker_pool := printf "projects/%s/locations/%s/workerPools/%s" .worker_pool.project .worker_pool.location .worker_pool.name}}
```
Comment (xingao267) [inline templates/tfengine/components/cicd/triggers.tf:28]: need to move to inside of `{{range get . "envs" -}}` so per-env can be achieved

File: templates/tfengine/components/cicd/triggers.tf (hunk 1 above)
Comment (pasha-gh) [reply templates/tfengine/components/cicd/triggers.tf:72]: done

File: templates/tfengine/components/cicd/triggers.tf (hunk 2 above)
Comment (pasha-gh) [reply templates/tfengine/components/cicd/triggers.tf]: done

File: docs/tfengine/schemas/cicd.md [hunk 3]
```diff
@@ -26,6 +26,10 @@
 | envs.triggers.validate | Config block for the presubmit validation Cloud Build trigger. If specified, create the trigger and grant the Cloud Build Service Account necessary permissions to perform the build. | object | false | - | - |
 | envs.triggers.validate.run_on_push | Whether or not to be automatically triggered from a PR/push to branch. Default to true. | boolean | false | - | - |
 | envs.triggers.validate.run_on_schedule | Whether or not to be automatically triggered according a specified schedule. The schedule is specified using [unix-cron format](https://cloud.google.com/scheduler/docs/configuring/cron-job-schedules#defining_the_job_schedule) at Eastern Standard Time (EST). Default to none. | string | false | - | - |
+| envs.worker_pool | Optional worker pool attribute. Required for CICD to work with private cluster endpoints. | object | false | - | - |
```
Comment (xingao267) [inline docs/tfengine/schemas/cicd.md:29]: can you change some example config to use this new feature? e.g. multi_envs.hcl, and team.hcl

File: templates/tfengine/recipes/cicd.hcl [hunk 4]
```diff
@@ -259,6 +259,32 @@ schema = {
               }
             }
           }
+          worker_pool = {
+            description          = <<EOF
+              Optional worker pool attribute. Required for CICD to work with private cluster endpoints.
```
Comment (xingao267) [inline templates/tfengine/recipes/cicd.hcl:264]: Make the comment a bit more general. For example:

Optional Cloud Build private worker pool configuration. Required for CICD to access resources in a private network, .e.g GKE clusters with a private endpoint.

File: templates/tfengine/recipes/cicd.hcl [hunk 5]
```diff
@@ -259,6 +259,32 @@ schema = {
...
+            type                 = "object"
+            additionalProperties = false
+            required = [
+              "project",
+              "location",
+              "name"
+            ]
+            properties = {
+              project = {
+                description = "The project worker pool belongs to."
```
Comment (xingao267) [inline templates/tfengine/recipes/cicd.hcl:275]: The project where the worker pool belongs.

File: docs/tfengine/schemas/cicd.md (hunk 3 above)
Comment (pasha-gh) [reply docs/tfengine/schemas/cicd.md]: done

File: templates/tfengine/recipes/cicd.hcl (hunk 5 above)
Comment (pasha-gh) [reply templates/tfengine/recipes/cicd.hcl:276]: done

File: templates/tfengine/recipes/cicd.hcl (hunk 4 above)
Comment (pasha-gh) [reply templates/tfengine/recipes/cicd.hcl]: done

File: templates/tfengine/components/cicd/configs/tf-apply.yaml [hunk 6]
```diff
@@ -38,3 +39,6 @@ steps:
     args: ["./cicd/configs/run.sh", "-d", "${_MANAGED_DIRS}", "-a", "init", "-a", "plan", "-a", "apply -auto-approve"]
     dir: "${_TERRAFORM_ROOT}"
     id: Apply
+
+options:
+  workerPool: "${_WORKER_POOL}"
```
Comment (MartinPetkov) [inline templates/tfengine/components/cicd/configs/tf-apply.yaml:44]: Shouldn't this just be `pool` and be an object? See the example here: https://cloud.google.com/build/docs/private-pools/run-builds-in-private-pool

Comment (pasha-gh) [reply templates/tfengine/components/cicd/configs/tf-apply.yaml:44]: I tested this. Using `pool: name:` and `workerPool` lead to the same results.

Comment (MartinPetkov) [reply templates/tfengine/components/cicd/configs/tf-apply.yaml:44]: Interesting. Could you find documentation for this variation then? I think if in the future someone Googles this they might end up on the same docs as me and wonder why it's working.

Comment (xingao267) [reply templates/tfengine/components/cicd/configs/tf-apply.yaml:44]: https://cloud.google.com/build/docs/api/reference/rest/v1/projects.builds#Build.BuildOptions

seems `workerPool` is a legacy field and is deprecated. `pool` is the new one.

Comment (MartinPetkov) [reply templates/tfengine/components/cicd/configs/tf-apply.yaml:44]: Ah so my suspicion was valid.

Comment (pasha-gh) [reply templates/tfengine/components/cicd/configs/tf-apply.yaml:44]: This terraform doc indicates `worker-pool` as a valid argument for `google_cloudbuild_trigger` `build-trigger`. However, it's doesn't specifically include the valid arguments for yaml files. I assume the arguments should have a parity.

Comment (pasha-gh) [reply templates/tfengine/components/cicd/configs/tf-apply.yaml:44]: Ah I see! Nice find!

Comment (xingao267) [reply templates/tfengine/components/cicd/configs/tf-apply.yaml:44]: The official gcp API doc should be treated as the SoT. Terraform resources fields/docs are derived from those and can be out-of-date more easily. The `workerPool` field in https://cloud.google.com/build/docs/api/reference/rest/v1/projects.builds#Build.BuildOptions is indicated as deprecated.

Comment (pasha-gh) [reply templates/tfengine/components/cicd/configs/tf-apply.yaml]: Updated the to `pool.name`

**Review by MartinPetkov — COMMENTED** [review]
LGTM but I'll let Xin approve.

**Review by xingao267 — APPROVED** [review]

**Pull request merged**